File: 0207-course-schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        premap = {i:[] for i in range(numCourses)}
        visited = set()
        for x in prerequisites:
            premap[x[0]].append(x[1])
        
        def dfs(cur):
            if cur in visited:
                return False
            if premap[cur] == []:
                return True
            visited.add(cur)
            for x in premap[cur]:
                if not dfs(x):
                    return False
            premap[cur] = []
            visited.remove(cur)
            return True 

        for x in range(numCourses):
            if not dfs(x):
                return False
        return True

File: 0207-course-schedule/solution.cpp
class Solution {
public:
    unordered_map<int, vector<int>> preqMap;
    //maps preq to a list of all courses that need that preq
    unordered_set<int> prevs;
    bool dfs(int curCourse){
        if(prevs.count(curCourse)){
            return false;
        }
        if(preqMap[curCourse].empty()){
            return true;
        }
        
        prevs.insert(curCourse);
        for(auto cur : preqMap[curCourse]){
            if(!dfs(cur)){
                return false;
            }
        }
        prevs.erase(curCourse);
        preqMap[curCourse].clear();
        return true;
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        //populate our preqMap
        for(int i = 0; i < numCourses; ++i){
            preqMap[i] = {};
        }
        for(auto cur : prerequisites){
            preqMap[cur[0]].push_back(cur[1]);
        }
        for(int i = 0; i < numCourses; ++i){
            if(!dfs(i)){
                return false;
            }
        }
        return true;
    }
};
